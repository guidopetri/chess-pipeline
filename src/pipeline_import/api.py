#! /usr/bin/env python3


def get_cloud_eval(fen, variation_count=1, variant='standard'):

    # ignore cloud eval for now because of API issues

    return {}
