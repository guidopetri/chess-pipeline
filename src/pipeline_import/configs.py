#! /usr/bin/env python3

import configparser


def _load_cfg():
    cfg = configparser.ConfigParser()
    cfg.read('/io/config.toml')
    return cfg


def get_cfg(key: str):
    return _load_cfg()[key]
