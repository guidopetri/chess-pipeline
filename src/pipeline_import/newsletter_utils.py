#! /usr/bin/env python3


def get_color_stats_text(color_stats):
    color_stats.reset_index(inplace=True, drop=False)

    category = color_stats['time_control_category'].head(1).item()
    results = color_stats[color_stats['time_control_category'] == category]
    first_results = results.head(1)
    first_color = first_results['player_color'].item()
    first_win_rate = first_results['Win'].item() * 100
    first_win_rate_str = (f'You had a {first_win_rate:.2f}% win rate'
                          f' with {first_color} in {category}')

    if color_stats.shape[0] > 1:
        if results.shape[0] > 1:
            results = results.tail(1)
            second_category_substr = ''
        else:
            # second row should be a different category
            results = color_stats.iloc[[1]]

            # results should be non-empty no matter what
            # since we have 2 rows at least, and either we have the same
            # category (we don't hit this line) or we have different
            # categories (this line is valid)
            second_category = results['time_control_category'].item()
            second_category_substr = ' in ' + second_category

        second_color = results['player_color'].item()
        second_win_rate = results['Win'].item() * 100

        second_win_rate_str = (f' and a {second_win_rate:.2f}% win rate '
                               f'with {second_color}'
                               f'{second_category_substr}')

    else:
        second_win_rate_str = ''

    win_rate_string = first_win_rate_str + second_win_rate_str + '.'

    return win_rate_string
