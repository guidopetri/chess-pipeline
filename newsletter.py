#! /usr/bin/env python3

from luigi import Task, LocalTarget
from luigi.format import Nop
from luigi.util import requires
from luigi.parameter import Parameter, ListParameter
from configs import sendgrid, newsletter_cfg, postgres_cfg


class GetData(Task):

    player = Parameter()
    columns = ListParameter(default=[])

    def run(self):
        from psycopg2 import connect
        from pandas import DataFrame

        pg_cfg = postgres_cfg()
        db_connection_string = 'postgresql://{}:{}@{}:{}/{}'

        with connect(db_connection_string.format(pg_cfg.read_user,
                                                 pg_cfg.read_password,
                                                 pg_cfg.host,
                                                 pg_cfg.port,
                                                 pg_cfg.database)) as con:
            cursor = con.cursor()

            sql = """SELECT * from chess_games
                     WHERE player = '{}'
                     AND datetime_played >= now()::date - interval '7 days';
                  """

            cursor.execute(sql.format(  # (', '.join(self.columns),
                                       self.player))
            colnames = [desc.name for desc in cursor.description]
            results = cursor.fetchall()

        df = DataFrame.from_records(results, columns=colnames)

        with self.output().temporary_path() as temp_output_path:
            df.to_pickle(temp_output_path, compression=None)

    def output(self):
        import os

        file_location = '~/Temp/luigi/week-data-{}.pckl'.format(self.player)
        return LocalTarget(os.path.expanduser(file_location), format=Nop)


@requires(GetData)
class WinRatioByColor(Task):

    def output(self):
        import os

        file_location = '~/Temp/luigi/graphs/win-by-color.pckl'
        return LocalTarget(os.path.expanduser(file_location), format=Nop)

    def run(self):
        import pickle
        import os
        from pandas import read_pickle
        from seaborn import set as sns_set

        with self.input().open('r') as f:
            df = read_pickle(f, compression=None)

        color_stats = df.groupby(['time_control_category',
                                  'player_color',
                                  'player_result']).agg({'id': 'nunique'})
        color_stats.reset_index(drop=False, inplace=True)

        # pivot so the columns are the player result
        color_stats = color_stats.pivot_table(index=['time_control_category',
                                                     'player_color'],
                                              columns='player_result',
                                              values='id')

        # divide each row by the sum of the row
        color_stats = color_stats.div(color_stats.sum(axis=1), axis=0)

        # set seaborn style for plots
        sns_set(style='whitegrid')

        # reorder columns
        color_stats = color_stats[['Win', 'Draw', 'Loss']]

        ax = color_stats.plot(kind='bar',
                              stacked=True,
                              color='gyr',
                              ylim=(0, 1),
                              rot=0,
                              title='Win-loss ratio by color played',
                              yticks=[0.0,
                                      0.1,
                                      0.2,
                                      0.3,
                                      0.4,
                                      0.5,
                                      0.6,
                                      0.7,
                                      0.8,
                                      0.9,
                                      1.0,
                                      1.01],  # enforce two digits of precision
                              )
        ax.set_ylabel('Ratio')
        ax.set_xlabel('Category / Color')
        ax.legend().set_title('')  # remove title

        for p in ax.patches:
            # place win% in the bar itself
            ax.annotate('{:.2f}%'.format(100 * p.get_height()),
                        xy=(0.5, 0.5),
                        xycoords=p,
                        ha='center',
                        va='center',
                        )

        # save the figure
        fig_loc = '~/Temp/luigi/graphs'
        fig_loc = os.path.expanduser(fig_loc)
        os.makedirs(fig_loc, exist_ok=True)
        ax.get_figure().savefig(os.path.join(fig_loc, 'win-by-color.png'),
                                bbox_inches='tight')

        text = ('You had a {:.2f}% win rate with {} in {}'
                ' and a {:.2f}% win rate with {}. <br>'
                '<img alt=\'Win rate by color played\' src='
                '\'cid:win-by-color\'><br>'
                .format(100 * color_stats.iloc[0][0],
                        *color_stats.iloc[0].name[::-1],
                        100 * color_stats.iloc[1][0],
                        color_stats.iloc[1].name[1],
                        ))

        with self.output().open('w') as f:
            pickle.dump(text, f, protocol=-1)


@requires(GetData)
class EloByWeekday(Task):

    def output(self):
        import os

        file_location = '~/Temp/luigi/graphs/elo-by-weekday.pckl'
        return LocalTarget(os.path.expanduser(file_location), format=Nop)

    def run(self):
        import pickle
        import os
        from pandas import read_pickle
        from seaborn import set as sns_set

        with self.input().open('r') as f:
            df = read_pickle(f, compression=None)

        df = df[df['time_control_category'] == 'blitz']
        df['weekday_played'] = df['datetime_played'].dt.weekday

        elo = (df.groupby('weekday_played')
                 .agg({'player_elo': ['mean',
                                      'std',
                                      'min',
                                      'max']}))
        # drop the first index on columns
        elo = (elo.T
                  .reset_index(level=0, drop=True)
                  .T
                  .reset_index(drop=False))

        sns_set(style='whitegrid')

        # plot main line with standard deviation
        ax = elo.plot(x='weekday_played',
                      y='mean',
                      yerr='std',
                      color='#0000FF',
                      title='Elo evolution by day of week',
                      style=[''],
                      legend=False,
                      capsize=4,
                      capthick=1,
                      )

        # plot min/maxes
        elo.plot(x='weekday_played',
                 y=['min', 'max'],
                 color='#999999',
                 style=['--', '--'],
                 ax=ax,
                 legend=False,
                 xlim=[-0.05, 6.05],
                 xticks=range(0, 7),
                 )

        min_last_day = elo[elo['weekday_played'] == 6]['min'].values
        max_last_day = elo[elo['weekday_played'] == 6]['max'].values
        mean_last_day = elo[elo['weekday_played'] == 6]['mean'].values

        # annotate the lines individually
        ax.annotate('min',
                    xy=(6.05, min_last_day),
                    color='#555555',
                    )
        ax.annotate('mean + std',
                    xy=(6.05, mean_last_day),
                    color='k',
                    )
        ax.annotate('max',
                    xy=(6.05, max_last_day),
                    color='#555555',
                    )

        # change the tick labels
        ax.set_xticklabels(['Monday',
                            'Tuesday',
                            'Wednesday',
                            'Thursday',
                            'Friday',
                            'Saturday',
                            'Sunday',
                            ],
                           rotation=45)

        ax.set_xlabel('Weekday')
        ax.set_ylabel('Rating')

        # save the figure
        fig_loc = '~/Temp/luigi/graphs'
        fig_loc = os.path.expanduser(fig_loc)
        os.makedirs(fig_loc, exist_ok=True)
        ax.get_figure().savefig(os.path.join(fig_loc, 'elo-by-weekday.png'),
                                bbox_inches='tight')

        max_elo = elo['max'].max()
        min_elo = elo['min'].min()

        text = ('This week, your highest elo in blitz was {}, and'
                ' your lowest elo was {}. <br>'
                '<img alt=\'Elo by weekday\' src='
                '\'cid:elo-by-weekday\'><br>'
                .format(max_elo,
                        min_elo,
                        ))

        with self.output().open('w') as f:
            pickle.dump(text, f, protocol=-1)


@requires(WinRatioByColor, EloByWeekday)
class CreateNewsletter(Task):

    receiver = Parameter()

    def run(self):
        import pickle
        import base64
        import os
        from sendgrid.helpers import mail
        from bs4 import BeautifulSoup

        newsletter = mail.Mail(from_email=newsletter_cfg().sender,
                               to_emails=self.receiver,
                               subject=('Chess Newsletter - {}'
                                        .format(self.player)),
                               )

        imgs_loc = os.path.expanduser('~/Temp/luigi/graphs/')

        for file in os.listdir(imgs_loc):
            if file.endswith('.png'):
                with open(os.path.join(imgs_loc, file), 'rb') as f:
                    encoded_img = base64.b64encode(f.read()).decode('utf-8')

                attachment = mail.Attachment(file_content=encoded_img,
                                             file_name=file,
                                             file_type='image/png',
                                             disposition='inline',
                                             content_id=file[:-4],
                                             )

                newsletter.add_attachment(attachment)

        message = ['<html><body> Hi {},<br><br>'
                   'This week you played chess! Here\'s your performance:'
                   .format(self.player)
                   ]

        for inp in self.input():
            with inp.open('r') as f:
                text = pickle.load(f)
                message.append(text)

        message.append('Hope you do well this upcoming week!</body></html>')

        full_message = '<br>'.join(message)

        newsletter.add_content(full_message, 'text/html')

        # add plaintext MIME to make it less likely to be categorized as spam
        newsletter.add_content(BeautifulSoup(full_message,
                                             'html.parser').get_text(),
                               'text/plain')

        with self.output().open('w') as f:
            pickle.dump(newsletter, f, protocol=-1)

    def output(self):
        import os

        file_location = '~/Temp/luigi/newsletter.pckl'
        return LocalTarget(os.path.expanduser(file_location), format=Nop)


@requires(CreateNewsletter)
class SendNewsletter(Task):

    result = False

    def run(self):
        import os
        import shutil
        import pickle
        from sendgrid import SendGridAPIClient

        with self.input().open('r') as f:
            newsletter = pickle.load(f)

        client = SendGridAPIClient(sendgrid().apikey)
        response = client.send(newsletter)

        self.result = response.status_code == 202

        filepath = os.path.expanduser('~/Temp/luigi')

        for file in os.listdir(filepath):
            full_path = os.path.join(filepath, file)
            if os.path.isfile(full_path):
                os.remove(full_path)
            elif os.path.isdir(full_path):
                shutil.rmtree(full_path)

    def complete(self):
        return self.result

    def output(self):
        pass
