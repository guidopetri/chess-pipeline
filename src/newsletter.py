#! /usr/bin/env python3

from luigi import Task, LocalTarget
from luigi.format import Nop
from luigi.util import requires
from luigi.parameter import Parameter, ListParameter
from pipeline_import.configs import sendgrid, newsletter_cfg, postgres_cfg
from pipeline_import.transforms import get_color_stats, get_elo_by_weekday
from pipeline_import.transforms import get_weekly_data
from pipeline_import.plots import make_color_stats_plot
from pipeline_import.plots import make_elo_by_weekday_plot


class GetData(Task):

    player = Parameter()
    columns = ListParameter(default=[])

    def run(self):
        pg_cfg = postgres_cfg()
        df = get_weekly_data(pg_cfg, self.player)

        with self.output().temporary_path() as temp_output_path:
            df.to_pickle(temp_output_path, compression=None)

    def output(self):
        import os

        file_location = f'~/Temp/luigi/week-data-{self.player}.pckl'
        return LocalTarget(os.path.expanduser(file_location), format=Nop)


@requires(GetData)
class WinRatioByColor(Task):

    def output(self):
        import os

        file_loc = f'~/Temp/luigi/graphs/win-by-color-{self.player}.pckl'
        return LocalTarget(os.path.expanduser(file_loc), format=Nop)

    def run(self):
        import pickle
        import os
        from pandas import read_pickle

        with self.input().open('r') as f:
            df = read_pickle(f, compression=None)

        if df.empty:
            text = 'Wait a second, no you didn\'t!'

            with self.output().open('w') as f:
                pickle.dump(text, f, protocol=-1)

            return

        color_stats = get_color_stats(df)

        fig_loc = '~/Temp/luigi/graphs'
        fig_loc = os.path.expanduser(fig_loc)
        filename = f'win-by-color-{self.player}.png'
        make_color_stats_plot(color_stats,
                              fig_loc=fig_loc,
                              filename=filename)

        color_stats.reset_index(inplace=True, drop=False)

        category = color_stats['time_control_category'].head(1).item()
        results = color_stats[color_stats['time_control_category'] == category]
        first_results = results.head(1)
        first_color = first_results['player_color'].item()
        first_win_rate = first_results['Win'].item() * 100
        first_win_rate_str = (f'You had a {first_win_rate:.2f}% win rate'
                              f' with {first_color} in {category}')

        if color_stats.shape[0] > 1:
            if results.shape[0] > 1:
                results = results.tail(1)
                second_category_substr = ''
            else:
                # second row should be a different category
                results = color_stats.iloc[[1]]

                # results should be non-empty no matter what
                # since we have 2 rows at least, and either we have the same
                # category (we don't hit this line) or we have different
                # categories (this line is valid)
                second_category = results['time_control_category'].item()
                second_category_substr = ' in ' + second_category

            second_color = results['player_color'].item()
            second_win_rate = results['Win'].item() * 100

            second_win_rate_str = (f' and a {second_win_rate:.2f}% win rate '
                                   f'with {second_color}'
                                   f'{second_category_substr}')

        else:
            second_win_rate_str = ''

        win_rate_string = first_win_rate_str + second_win_rate_str + '.'

        text = (win_rate_string
                + ' <br> <img alt=\'Win rate by color '
                + 'played\' src=\'cid:win-by-color\'><br>'
                )

        with self.output().open('w') as f:
            pickle.dump(text, f, protocol=-1)


@requires(GetData)
class EloByWeekday(Task):

    category = Parameter(default='blitz')

    def output(self):
        import os

        file_loc = f'~/Temp/luigi/graphs/elo-by-weekday-{self.player}.pckl'
        return LocalTarget(os.path.expanduser(file_loc), format=Nop)

    def run(self):
        import pickle
        import os
        from pandas import read_pickle

        with self.input().open('r') as f:
            df = read_pickle(f, compression=None)

        if df.empty:
            text = '\n'

            with self.output().open('w') as f:
                pickle.dump(text, f, protocol=-1)

            return

        elo = get_elo_by_weekday(df, category=self.category)

        fig_loc = '~/Temp/luigi/graphs'
        fig_loc = os.path.expanduser(fig_loc)
        filename = f'elo-by-weekday-{self.player}.png'
        make_elo_by_weekday_plot(elo, fig_loc=fig_loc, filename=filename)

        max_elo = int(elo['max'].max())
        min_elo = int(elo['min'].min())

        text = (f'This week, your highest elo in {self.category} was '
                f'{max_elo} and your lowest elo was {min_elo}. <br>'
                f'<img alt=\'Elo by weekday\' src='
                f'\'cid:elo-by-weekday\'><br>'
                )

        with self.output().open('w') as f:
            pickle.dump(text, f, protocol=-1)


@requires(WinRatioByColor, EloByWeekday)
class CreateNewsletter(Task):

    receiver = Parameter()

    def run(self):
        import pickle
        import base64
        import os
        from sendgrid.helpers import mail
        from bs4 import BeautifulSoup

        newsletter = mail.Mail(from_email=newsletter_cfg().sender,
                               to_emails=self.receiver,
                               subject=f'Chess Newsletter - {self.player}',
                               )

        imgs_loc = os.path.expanduser('~/Temp/luigi/graphs/')

        for file in os.listdir(imgs_loc):
            if file.endswith('.png') and self.player in file:
                with open(os.path.join(imgs_loc, file), 'rb') as f:
                    encoded_img = base64.b64encode(f.read()).decode('utf-8')

                attachment = mail.Attachment(file_content=encoded_img,
                                             file_name=file,
                                             file_type='image/png',
                                             disposition='inline',
                                             content_id=file[:-4],
                                             )

                newsletter.add_attachment(attachment)

        message = [f'<html><body> Hi {self.player},<br><br>'
                   f'This week you played chess! Here\'s your performance:'
                   ]

        for inp in self.input():
            with inp.open('r') as f:
                text = pickle.load(f)
                message.append(text)

        message.append('Hope you do well this upcoming week!</body></html>')

        full_message = '<br>'.join(message)

        newsletter.add_content(full_message, 'text/html')

        # add plaintext MIME to make it less likely to be categorized as spam
        newsletter.add_content(BeautifulSoup(full_message,
                                             'html.parser').get_text(),
                               'text/plain')

        with self.output().open('w') as f:
            pickle.dump(newsletter, f, protocol=-1)

    def output(self):
        import os

        file_loc = f'~/Temp/luigi/newsletter-{self.player}.pckl'
        return LocalTarget(os.path.expanduser(file_loc), format=Nop)


@requires(CreateNewsletter)
class SendNewsletter(Task):

    result = False

    def run(self):
        import os
        import shutil
        import pickle
        from sendgrid import SendGridAPIClient

        with self.input().open('r') as f:
            newsletter = pickle.load(f)

        client = SendGridAPIClient(sendgrid().apikey)
        response = client.send(newsletter)

        self.result = response.status_code == 202

        filepath = os.path.expanduser('~/Temp/luigi')

        for file in os.listdir(filepath):
            full_path = os.path.join(filepath, file)
            if os.path.isfile(full_path):
                os.remove(full_path)
            elif os.path.isdir(full_path):
                shutil.rmtree(full_path)

    def complete(self):
        return self.result

    def output(self):
        pass
