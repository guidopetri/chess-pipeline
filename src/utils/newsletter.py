import base64
import os
import shutil

from bs4 import BeautifulSoup
from pipeline_import.configs import get_cfg
from pipeline_import.plots import (
    make_color_stats_plot,
    make_elo_by_weekday_plot,
)
from pipeline_import.transforms import (
    get_color_stats,
    get_elo_by_weekday,
)
from sendgrid import SendGridAPIClient
from sendgrid.helpers import mail


def get_color_stats_text(color_stats):
    color_stats.reset_index(inplace=True, drop=False)

    category = color_stats['time_control_category'].head(1).item()
    results = color_stats[color_stats['time_control_category'] == category]
    first_results = results.head(1)
    first_color = first_results['player_color'].item()
    first_win_rate = first_results['Win'].item() * 100
    first_win_rate_str = (f'You had a {first_win_rate:.2f}% win rate'
                          f' with {first_color} in {category}')

    if color_stats.shape[0] > 1:
        if results.shape[0] > 1:
            results = results.tail(1)
            second_category_substr = ''
        else:
            # second row should be a different category
            results = color_stats.iloc[[1]]

            # results should be non-empty no matter what
            # since we have 2 rows at least, and either we have the same
            # category (we don't hit this line) or we have different
            # categories (this line is valid)
            second_category = results['time_control_category'].item()
            second_category_substr = ' in ' + second_category

        second_color = results['player_color'].item()
        second_win_rate = results['Win'].item() * 100

        second_win_rate_str = (f' and a {second_win_rate:.2f}% win rate '
                               f'with {second_color}'
                               f'{second_category_substr}')

    else:
        second_win_rate_str = ''

    win_rate_string = first_win_rate_str + second_win_rate_str + '.'

    return win_rate_string


def generate_elo_by_weekday_text(df, category, player) -> str:
    if df.empty:
        return '\n'

    elo = get_elo_by_weekday(df, category=category)

    fig_loc = '~/Temp/luigi/graphs'
    fig_loc = os.path.expanduser(fig_loc)
    filename = f'elo-by-weekday-{player}.png'
    make_elo_by_weekday_plot(elo, fig_loc=fig_loc, filename=filename)

    max_elo = int(elo['max'].max())
    min_elo = int(elo['min'].min())

    text = (f'This week, your highest elo in {category} was '
            f'{max_elo} and your lowest elo was {min_elo}. <br>'
            f"<img alt='Elo by weekday' src="
            f"'cid:elo-by-weekday'><br>"
            )
    return text


def generate_win_ratio_by_color_text(df, player) -> str:
    if df.empty:
        return "Wait a second, no you didn't!"

    color_stats = get_color_stats(df)

    fig_loc = '~/Temp/luigi/graphs'
    fig_loc = os.path.expanduser(fig_loc)
    filename = f'win-by-color-{player}.png'
    make_color_stats_plot(color_stats,
                          fig_loc=fig_loc,
                          filename=filename)

    win_rate_string = get_color_stats_text(color_stats)

    text = (win_rate_string
            + " <br> <img alt='Win rate by color "
            + "played' src='cid:win-by-color'><br>"
            )
    return text


def send_newsletter(newsletter) -> bool:
    client = SendGridAPIClient(get_cfg('sendgrid')['apikey'])
    response = client.send(newsletter)

    filepath = os.path.expanduser('~/Temp/luigi')

    if os.path.exists(filepath):
        for file in os.listdir(filepath):
            full_path = os.path.join(filepath, file)
            if os.path.isfile(full_path):
                os.remove(full_path)
            elif os.path.isdir(full_path):
                shutil.rmtree(full_path)
    return response.status_code == 202


def create_newsletter(texts, player, receiver):
    newsletter = mail.Mail(from_email=get_cfg('newsletter_cfg')['sender'],
                           to_emails=receiver,
                           subject=f'Chess Newsletter - {player}',
                           )

    imgs_loc = os.path.expanduser('~/Temp/luigi/graphs/')

    for file in os.listdir(imgs_loc):
        if file.endswith('.png') and player in file:
            with open(os.path.join(imgs_loc, file), 'rb') as f:
                encoded_img = base64.b64encode(f.read()).decode('utf-8')

            attachment = mail.Attachment(file_content=encoded_img,
                                         file_name=file,
                                         file_type='image/png',
                                         disposition='inline',
                                         content_id=file[:-4],
                                         )

            newsletter.add_attachment(attachment)

    message = [f'<html><body> Hi {player},<br><br>'
               f"This week you played chess! Here's your performance:"
               ]

    message.extend(texts)

    message.append('Hope you do well this upcoming week!</body></html>')

    full_message = '<br>'.join(message)

    newsletter.add_content(full_message, 'text/html')

    # add plaintext MIME to make it less likely to be categorized as spam
    newsletter.add_content(BeautifulSoup(full_message,
                                         'html.parser').get_text(),
                           'text/plain')
    return newsletter
