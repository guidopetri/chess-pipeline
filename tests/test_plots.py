#! /usr/bin/env python3

import contextlib
import hashlib
import os
import pickle
from io import BytesIO

import pandas as pd
from pipeline_import import plots
from utils.newsletter import (
    create_newsletter,
    generate_elo_by_weekday_text,
    generate_win_ratio_by_color_text,
    get_color_stats_text,
    send_newsletter,
)


def test_color_stats_plot(snapshot):

    multiindex = pd.MultiIndex.from_arrays([['blitz', 'blitz', 'bullet'],
                                            ['white', 'black', 'black']],
                                           names=('time_control_category',
                                                  'player_color'))

    color_stats = pd.DataFrame([[1 / 3, 1 / 3, 1 / 3],
                                [1, 0, 0],
                                [1, 0, 0]],
                               columns=['Win', 'Draw', 'Loss'],
                               index=multiindex,
                               )

    fig_loc = '.'
    filename = 'color_stats_test.png'

    plots.make_color_stats_plot(color_stats, fig_loc, filename)

    file_loc = os.path.join(fig_loc, filename)

    with open(file_loc, 'rb') as f:
        md5 = hashlib.md5(f.read()).hexdigest()

    assert md5 == snapshot

    os.remove(file_loc)


def test_elo_by_weekday_plot(snapshot):
    elo = pd.DataFrame([[0, 1666, 0, 1666.0, 1666.0],
                        [1, 1685, 7.071, 1680.0, 1690.0],
                        [2, 1685, 7.071, 1680.0, 1690.0],
                        [3, 1685, 7.071, 1680.0, 1690.0],
                        [4, 1685, 7.071, 1680.0, 1690.0],
                        [5, 1685, 7.071, 1680.0, 1690.0],
                        [6, 1662.5, 3.536, 1660.0, 1665.0]],
                       columns=['weekday_played',
                                'mean',
                                'std',
                                'min',
                                'max'],
                       )

    fig_loc = '.'
    filename = 'elo_by_weekday_test.png'

    plots.make_elo_by_weekday_plot(elo, fig_loc, filename)

    file_loc = os.path.join(fig_loc, filename)

    with open(file_loc, 'rb') as f:
        md5 = hashlib.md5(f.read()).hexdigest()

    assert md5 == snapshot

    os.remove(file_loc)


def test_elo_by_weekday_without_games(snapshot):

    empty_elo = pd.DataFrame([],
                             columns=['weekday_played',
                                      'mean',
                                      'std',
                                      'min',
                                      'max'],
                             # enforce float dtypes,
                             # otherwise the error doesn't show up
                             dtype=float,
                             )

    fig_loc = '.'
    filename = 'elo_by_weekday_without_games.png'

    plots.make_elo_by_weekday_plot(empty_elo, fig_loc, filename)

    file_loc = os.path.join(fig_loc, filename)

    with open(file_loc, 'rb') as f:
        md5 = hashlib.md5(f.read()).hexdigest()

    assert md5 == snapshot

    os.remove(file_loc)


def test_color_stats_text_generic():

    multiindex = pd.MultiIndex.from_arrays([['blitz', 'blitz', 'bullet'],
                                            ['white', 'black', 'black']],
                                           names=('time_control_category',
                                                  'player_color'))

    color_stats = pd.DataFrame([[1 / 3, 1 / 3, 1 / 3],
                                [1, 0, 0],
                                [1, 0, 0]],
                               columns=['Win', 'Draw', 'Loss'],
                               index=multiindex,
                               )

    win_rate_str = get_color_stats_text(color_stats)

    true_win_rate_str = ('You had a 33.33% win rate with white in blitz and'
                         ' a 100.00% win rate with black.')

    assert win_rate_str == true_win_rate_str


def test_color_stats_text_only_one_color():

    multiindex = pd.MultiIndex.from_arrays([['blitz'],
                                            ['black']],
                                           names=('time_control_category',
                                                  'player_color'))

    color_stats = pd.DataFrame([[1 / 3, 1 / 3, 1 / 3]],
                               columns=['Win', 'Draw', 'Loss'],
                               index=multiindex,
                               )

    win_rate_str = get_color_stats_text(color_stats)

    true_win_rate_str = 'You had a 33.33% win rate with black in blitz.'

    assert win_rate_str == true_win_rate_str


def test_color_stats_text_multiple_categories():

    multiindex = pd.MultiIndex.from_arrays([['blitz', 'bullet'],
                                            ['black', 'black']],
                                           names=('time_control_category',
                                                  'player_color'))

    color_stats = pd.DataFrame([[1 / 3, 1 / 3, 1 / 3],
                                [1, 0, 0]],
                               columns=['Win', 'Draw', 'Loss'],
                               index=multiindex,
                               )

    win_rate_str = get_color_stats_text(color_stats)

    true_win_rate_str = ('You had a 33.33% win rate with black in blitz and'
                         ' a 100.00% win rate with black in bullet.')

    assert win_rate_str == true_win_rate_str


def test_generate_elo_by_weekday_text_empty():
    input_df = pd.DataFrame()
    text = generate_elo_by_weekday_text(input_df, '', '')

    assert text == '\n'


def test_generate_elo_by_weekday_text_generic(mocker, snapshot):
    # TODO: split this function off so it doesn't have to be patched
    mocker.patch('utils.newsletter.make_elo_by_weekday_plot')

    elo_df = pd.DataFrame([[100, 200], [200, 300]], columns=['min', 'max'])

    mocker.patch('utils.newsletter.get_elo_by_weekday', return_value=elo_df)

    text = generate_elo_by_weekday_text(elo_df, 'bullet', 'thibault')

    assert text == snapshot


def test_generate_win_ratio_by_color_text_empty(snapshot):
    input_df = pd.DataFrame()
    text = generate_win_ratio_by_color_text(input_df, '')

    assert text == snapshot


def test_generate_win_ratio_by_color_text_generic(mocker, snapshot):
    mocker.patch('utils.newsletter.get_color_stats')
    mocker.patch('utils.newsletter.get_color_stats_text', return_value='foo')
    # TODO: split this function off so it doesn't have to be patched
    mocker.patch('utils.newsletter.make_color_stats_plot')
    text = generate_win_ratio_by_color_text(pd.DataFrame([0]), '')

    assert text == snapshot


def test_send_newsletter(tmp_path, mocker):
    mocker.patch('utils.newsletter.get_cfg')
    mock_send = mocker.patch('utils.newsletter.SendGridAPIClient.send')
    mock_send.return_value.status_code = 202

    mocker.patch('os.path.expanduser', return_value=tmp_path / 'files')
    os.makedirs(tmp_path / 'files' / 'foobar')

    with open(tmp_path / 'files' / 'foo.bar', 'w') as f:
        f.write('dummy-input')

    text = 'foo'

    assert send_newsletter(text)
    mock_send.assert_called_once_with(text)
    assert not list((tmp_path / 'files').glob('*'))


def test_create_newsletter(tmp_path, mocker, snapshot):
    mocker.patch('utils.newsletter.get_cfg',
                 return_value={'sender': 'foo@bar.com'},
                 )

    mocker.patch('os.path.expanduser', return_value=tmp_path / 'images')
    os.mkdir(tmp_path / 'images')

    with open(tmp_path / 'images' / 'thibault.png', 'w') as f:
        f.write('dummy-input')

    class FakeLuigiInput:
        def __init__(self, value):
            self.value = value

        @contextlib.contextmanager
        def open(self, *args):
            return (x for x in [self.value])

    inputs = [FakeLuigiInput(BytesIO(pickle.dumps('foo'))),
              FakeLuigiInput(BytesIO(pickle.dumps('bar'))),
              ]

    newsletter = create_newsletter(inputs, 'thibault', 'bar@foo.com')

    assert newsletter == snapshot
