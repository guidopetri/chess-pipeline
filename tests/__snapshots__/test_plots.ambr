# serializer version: 1
# name: test_color_stats_plot
  '6f3cbda843e6997c51f6cd398995449c'
# ---
# name: test_create_newsletter
  Mail(
    asm=None,
    attachment=None,
    attachments=list([
      Attachment(
        content_id=ContentId(
          content_id='thibault',
        ),
        disposition=Disposition(
          disposition='inline',
        ),
        file_content=FileContent(
          file_content='ZHVtbXktaW5wdXQ=',
        ),
        file_name=FileName(
          file_name='thibault.png',
        ),
        file_type=FileType(
          file_type='image/png',
        ),
      ),
    ]),
    batch_id=None,
    bcc=None,
    categories=None,
    category=None,
    cc=None,
    content=None,
    contents=list([
      Content(
        content=" Hi thibault,This week you played chess! Here's your performance:foobarHope you do well this upcoming week!",
        mime_type='text/plain',
      ),
      Content(
        content="<html><body> Hi thibault,<br><br>This week you played chess! Here's your performance:<br>foo<br>bar<br>Hope you do well this upcoming week!</body></html>",
        mime_type='text/html',
      ),
    ]),
    custom_arg=None,
    custom_args=None,
    dynamic_template_data=None,
    from_email=From(
      email='foo@bar.com',
      name=None,
      personalization=0,
      subject=None,
      substitutions=None,
    ),
    header=None,
    headers=None,
    ip_pool_name=None,
    mail_settings=None,
    personalizations=list([
      Personalization(
        bccs=list([
        ]),
        ccs=list([
        ]),
        custom_args=list([
        ]),
        dynamic_template_data=None,
        headers=list([
        ]),
        send_at=None,
        subject=None,
        substitutions=list([
        ]),
        tos=list([
          dict({
            'email': 'bar@foo.com',
          }),
        ]),
      ),
    ]),
    reply_to=None,
    section=None,
    sections=None,
    send_at=None,
    subject=Subject(
      personalization=None,
      subject='Chess Newsletter - thibault',
    ),
    substitution=None,
    template_id=None,
    to=None,
    tracking_settings=None,
  )
# ---
# name: test_elo_by_weekday_plot
  '8dafbe932ae3bd00255b68be55f418f3'
# ---
# name: test_elo_by_weekday_without_games
  'bf002fb2bcdf486bca20093e1e5e2f35'
# ---
# name: test_generate_elo_by_weekday_text_generic
  "This week, your highest elo in bullet was 300 and your lowest elo was 100. <br><img alt='Elo by weekday' src='cid:elo-by-weekday'><br>"
# ---
# name: test_generate_win_ratio_by_color_text_empty
  "Wait a second, no you didn't!"
# ---
# name: test_generate_win_ratio_by_color_text_generic
  "foo <br> <img alt='Win rate by color played' src='cid:win-by-color'><br>"
# ---
